### ISSUE TITLE
[HW2] resolve_structs creating duplicate types

### ISSUE THREAD

**Issue by KunJeong** [kaist-cp/cs420#280]
Hi. I'm currently debugging my IR for `examples/c/struct2.c`, and I've come across a problem that I can't figure out.

```c
// struct2.c
typedef struct {
  char a;
  struct {
    int b[4];
  };
  long c;
} Temp; //-1-

int main() {
    const Temp temp = {1, {{2, 3, 4, 5}}, 6};
    
    Temp temp2; //-2-
    temp2 = temp;

    int sum = temp2.a + temp2.b[2] + temp2.c;
    
    return sum == 11;
}
```
After the "1" part of the code, the `structs` hashmap is filled in correctly with two structs - `%t0` for the outer struct and `%t1` for the inner struct. (at `translate_decl`)
Also, the `typedefs` has one entry, "Temp", that has the corresponding struct type but with the name as `None`. (at `add_declaration`) (#)
(#) is problematic because at "2" part of the code, resolving the typedef "Temp" gives the struct type with name as `None`. Calling `resolve_structs` with this dtype results in creating `%t2` and `%t3`, which are duplicate versions of `%t0` and `%t1` with the difference that the name is `None`.

To fix the issue at (#), I thought that I should call `resolve_structs` at `add_declaration` as well, so that the name is not `None`. However, there is no `tempid_counter` in this scope (IRgen), and I think it doesn't make sense to have it.

On the other hand, if the behavior at (#) is intended, and I should search for the name afterwards, this would mean that I should find the struct definition with only the information `fields`, since everything else is `None`. However, I think this would be problematic as there could be multiple structs with the same fields but different names.

Am I missing something here? Thank you.

**Comment by cmpark0126** [comment]
Thank you for the question.

I think you used `Dtype::resolve_structs` method to fill `structs` hashmap. Is it right? 

If so, you can get `struct %t1` from the above method, and you can link this struct type to the typedef name `Temp`.

If you follow the step above, I believe you could obtain the struct type whose name is `%t1` at (2) after resolving the typedef `Temp`.

Perhaps the above answer can solve the problem.

**Comment by KunJeong** [comment]
Sorry, but I'm still confused.
> I think you used `Dtype::resolve_structs` method to fill `structs` hashmap. Is it right?

Yes. But I called resolve_structs after resolve_typedefs, because I got the error "typedef should be replaced with actual type" when I did it the other way around. Could this be the source of the problem that `typedefs` does not contain the struct name?
> If so, you can get struct %t1 from the above method, and you can link this struct type to the typedef name Temp.
> If you follow the step above, I believe you could obtain the struct type whose name is `%t1` at (2) after resolving the typedef `Temp`.

My problem was that calling resolve_structs at (2) already created the duplicates. So it returns the newly created `struct %t3` instead. Should I insert the name inside the `typedefs` hashmap manually? I guess it could be done, but I'm not sure it's recommended because of this explanation at [Structs High-level Explanation](https://github.com/kaist-cp/cs420/issues/256)
> All of the method or function related to struct type requires structs hash table as a function parameter because kecc assumes that all definitions of structs are only in the structs hash table.

Thank you for your help.

**Comment by cmpark0126** [comment]
Actually, I think you misunderstand the function of `resolve_typedefs` method. `resolve_typedefs` is a function that replaces the typedef name to actual type using `typedefs` Hashmap. However, this function cannot link the typedef name and its actual type, so you need to link `Temp` and `struct %t1` manually as you said.

This series of processes do not violate the above-mentioned explanation of Structs High-level Explanation. Because these processes are not part of `struct` type but `typedef` translation.

**Comment by KunJeong** [comment]
I see. So if I'm understanding correctly, the manual linking should be done at (1) and not (2), right?

**Comment by cmpark0126** [comment]
Exactly :)

**Comment by KunJeong** [comment]
Great! Thank you again XD

**Comment by KunJeong** [comment]
Hi. I've implemented the manual linking as you suggested, and I'm getting multiple problems.

1) I had to make `typedefs` mutable, unlike the videos, which wasn't a big problem but felt wrong.

2) Next, I got `redefinition of %t1` error, which was from `resolve_structs()` at `translate_decl`. It seems that unlike `resolve_typedefs()`, `resolve_structs()` only supports new insertions and throws errors at existing names. This is weird, but I fixed my code to not call `resolve_structs` for names already in the `struct` hashmap.

3) Now, sadly, I'm getting both `%t1` and `%t1:<a:i8, %anon:struct %anon:<b:[4 x i32]>, c:i64>`in my IR, and my IR keeps typecasting between the two types. I guess it means I should manually link inner structs (`%t0` in this case) as well. So the linking should be done recursively, which is possible but requires a lot more work.

The biggest problem is that this feels like a very inefficient solution to a simple problem. Mutating typedefs, manually linking, this all seems like something that resolve_typedefs and resolve_structs should take care of. Sorry to bother you with the same question over and over, but I'm really unsure whether I'm on the right path. Could you please confirm that I'm doing things correctly?

**Comment by cmpark0126** [comment]
Here are the answers to your questions.

1. We define `typedef` as a field of `Irgen`. So, we can use it as mutable in some methods of `Irgen`.

2. Actually, you don't need to use `resolve_structs ` at `translate_decl`. Please refer to [this link](https://github.com/kaist-cp/kecc-public/blob/56aabbe35d9a726e09b2a6b113a22d4743a665f6/src/c/parse.rs#L314-L322)

3. If the implementation is right, you don't need to consider `%t0` by yourself. Because `%t0` is used only at `%t1`. By the way, I don't understand why you think you need to link inner struct (`%t0`) manually. Could you try to compare your IR with `examples/ir0/struct2.ir`?

I cannot sure if these answers will help. If you have more questions, please comment.

**Comment by KunJeong** [comment]
> Actually, you don't need to use `resolve_structs` at `translate_decl`. Please refer to this link

I guess this is the main problem then. I originally tried to call `resolve_structs` at `add_declaration` because of this, but realized that `tempid_counter`, which is needed to call `resolve_structs`, didn't exists in `Irgen`. So I called at `translate_decl` instead, using the typedefs field of `IrgenFunc` (which is why I needed to "change" to mutable). I guess this was the source of this mess that I've put myself in.

Now I see that I could have  just given `mut usize 0` as the value for `tempid_counter`. I just tried this and everything seems to work fine. Thank you for your help.
